### PR TITLE
feat: Edit spending cap for ERC20 `approve` and `increaseAllowance`

### PR DESCRIPTION
**Pull request by pedronfigueiredo**
## **Description**

This PR adds a modal for editing the spending cap. It also refactors the `approve` component and adds integration tests for increaseAllowance.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26845?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/3078

## **Manual testing steps**

1. Go to this page...
2.
3.

### **After**

Screenshot 2024-09-03 at 14 33 08
Screenshot 2024-09-03 at 14 32 27

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/app/confirm/info/row/text.tsx [hunk 1]
```diff
@@ -37,6 +38,7 @@ export const ConfirmInfoRowText: React.FC<ConfirmInfoRowTextProps> = ({
   editIconClassName,
   tooltip,
   'data-testid': dataTestId,
+  editIconDataTestId,
```
Comment (jpuri) [inline ui/components/app/confirm/info/row/text.tsx:41]: may be use `"edit-nonce-icon"` as default value

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/__snapshots__/edit-spending-cap-modal.test.tsx.snap [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`<EditSpendingCapModal /> renders component 1`] = `<div />`;
```
Comment (jpuri) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/__snapshots__/edit-spending-cap-modal.test.tsx.snap:3]: Looks like component is empty here

**Review by jpuri — DISMISSED** [review]
LGTM

File: ui/pages/confirmations/components/confirm/info/approve/approve.test.tsx [hunk 3]
```diff
@@ -24,11 +23,51 @@ jest.mock(
...
+        },
+      ],
+    },
+    pending: false,
+  })),
+}));
+
 describe('<ApproveInfo />', () => {
   const middleware = [thunk];
+  useDecodedTransactionData;
```
Comment (digiwand) [inline ui/pages/confirmations/components/confirm/info/approve/approve.test.tsx:66]:
```suggestion
```

I think we can remove this

File: ui/pages/confirmations/components/confirm/info/approve/approve-static-simulation/approve-static-simulation.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    transactionMeta?.txParams?.to,
+    transactionMeta?.txParams?.from,
+    transactionMeta?.txParams?.data,
+  );
+
+  // decimals are sometimes undefined for a split second, so we need to ensure
+  // the spending cap is recalculated when the value updates
+  const [decimals, setDecimals] = useState(initialDecimals || '0');
+
+  useEffect(() => {
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/approve-static-simulation/approve-static-simulation.tsx:43]: I'm not sure what purpose this `useEffect` serves and why we need a `useState`?

Could we just set decimals as a standard variable based on `initialDecimals`?

File: ui/pages/confirmations/components/confirm/info/approve/approve.tsx [hunk 5]
```diff
@@ -1,223 +1,89 @@
...
-  const { pending } = useApproveTokenSimulation(
+  const customTxParamsData = useMemo(() => {
+    return getCustomTxParamsData(transactionMeta?.txParams?.data, {
+      customPermissionAmount: customSpendingCap || '0',
+      decimals,
+    });
+  }, [customSpendingCap, transactionMeta?.txParams?.data, decimals]);
+
+  const { value: estimatedGasLimit } = useAsyncResult(async () => {
+    return await estimateGas({
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/approve.tsx:51]: Gas is a scary one that got really complex in the legacy confirmations, could we encapsulate this behind a `useTransactionGasEstimate` so there's a single definition of how we calculate the gas for the current transaction?

File: ui/pages/confirmations/components/confirm/info/approve/approve.tsx [hunk 6]
```diff
@@ -1,223 +1,89 @@
...
+  }, [customTxParamsData, customSpendingCap, decimals]);
+
+  const [shouldUpdateConfirmation, setShouldUpdateConfirmation] =
+    useState(false);
+  const updateConfirmation = useCallback(() => {
+    if (shouldUpdateConfirmation && estimatedGasLimit) {
+      transactionMeta.txParams.data = customTxParamsData;
+      transactionMeta.txParams.gas = hexToDecimal(estimatedGasLimit as string);
+
+      dispatch(updateCurrentConfirmation(transactionMeta));
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/approve.tsx:66]: Are we not deriving the confirmation now via the context? Meaning the Redux state is no longer used?

File: ui/pages/confirmations/components/confirm/info/approve/approve.tsx [hunk 7]
```diff
@@ -227,9 +93,20 @@ const ApproveInfo = () => {
...
+          customSpendingCap={customSpendingCap}
+        />
+      )}
       <GasFeesSection />
       {showAdvancedDetails && <AdvancedDetails />}
+      <EditSpendingCapModal
+        isOpenEditSpendingCapModal={isOpenEditSpendingCapModal}
+        setIsOpenEditSpendingCapModal={setIsOpenEditSpendingCapModal}
+        customSpendingCap={customSpendingCap}
+        setCustomSpendingCap={setCustomSpendingCapCandidate}
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/approve.tsx:108]: Rather than adding further logic and complexity here, could we empower the modal to encapsulate the update logic and update the actual transaction?

File: ui/pages/confirmations/components/confirm/info/approve/approve.tsx [hunk 8]
```diff
@@ -1,223 +1,89 @@
...
-    decimals || '0',
-  );
+    dispatch,
+  ]);
 
-  if (pending) {
-    return <Container isLoading />;
-  }
+  useEffect(() => {
+    updateConfirmation();
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/approve.tsx:78]: Not sure why we have a callback that is only called from an effect? Would it be more readable to just include the logic directly within the effect?

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+  const { formattedSpendingCap } = useApproveTokenSimulation(
+    transactionMeta,
+    decimals || '0',
+  );
+
+  const [customSpendingCapCandidate, setCustomSpendingCapCandidate] =
+    useState('');
+
+  const onCancelHandler = () => {
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:66]: Minor, usual format is `handleX` so `handleCancel`.

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+          >
+            {t('editSpendingCapAccountBalance', [
+              accountBalance,
+              tokenSymbol || '',
+            ])}
+          </Text>
+        </ModalBody>
+        <ModalFooter
+          onSubmit={() => submitEditSpendingCap()}
+          onCancel={() => onCancelHandler()}
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:132]: Minor, there's no custom arguments needed here and they're already assigned to variables so no need for `() =>` here.

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: ui/pages/confirmations/components/confirm/info/approve/approve.tsx (hunk 6 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/approve.tsx:66]: That's true but we are currently in an "in between" state, I believe the changes to the submission flow will be done in [one of the next PRs](https://github.com/MetaMask/metamask-extension/pull/26685/files#diff-4935af77d52052112ed0a8efa32c2b8f63ceeb55cfa0704eb27ca3878033ccf7L101)

File: ui/pages/confirmations/components/confirm/info/approve/approve.tsx (hunk 7 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/approve.tsx:108]: The issue is that the spending cap needs to be shown in the `<SpendingCap />` component above it, so a variable needs to exist upstream from the modal component, in this <Approve /> section.

File: ui/pages/confirmations/components/confirm/info/approve/approve.tsx (hunk 6 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/approve.tsx]: I updated this now, as per https://github.com/MetaMask/metamask-extension/pull/26906#discussion_r1745111513

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/__snapshots__/edit-spending-cap-modal.test.tsx.snap (hunk 2 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/__snapshots__/edit-spending-cap-modal.test.tsx.snap:3]: There was an issue with rendering the portal. Mocking it fixed this.

**Review by jpuri — DISMISSED** [review]

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  const accountBalance = getAccountBalance(userBalance || '0', decimals || '0');
+
+  const { formattedSpendingCap } = useApproveTokenSimulation(
+    transactionMeta,
+    decimals || '0',
+  );
+
+  const [customSpendingCapInputValue, setCustomSpendingCapInputValue] =
+    useState('');
+  const [customSpendingCap, setCustomSpendingCap] = useState('');
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:68]: I'm confused by these two pieces of state.

The text box updates `customSpendingCapInputValue` but `customSpendingCap` is only set when submit is clicked?

If that's the case, would it not be simpler to only keep `customSpendingCapInputValue` as state.

And then do any validation and `dispatch` imperatively inside the `handleSubmit` rather than during render via a `useEffect`?

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    transactionMeta,
+    dispatch,
+  ]);
+
+  const handleCancel = () => {
+    setIsOpenEditSpendingCapModal(false);
+    setCustomSpendingCapInputValue('');
+  };
+
+  function handleSubmit() {
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:110]: Minor, `useCallback` for these handlers?

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    setIsOpenEditSpendingCapModal(false);
+    setCustomSpendingCapInputValue('');
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:112]: Does `handleSubmit` automatically close the modal as part of the modal component itself?

In that case, can we skip clearing the state and just validate and `dispatch`?

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+            {t('editSpendingCapDesc')}
+          </Text>
+          <TextField
+            type={TextFieldType.Number}
+            value={customSpendingCapInputValue}
+            onChange={(event) =>
+              setCustomSpendingCapInputValue(event.target.value)
+            }
+            placeholder={`${
+              customSpendingCap === ''
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:151]: Will `customSpendingCap` always be empty as it's only set on submit which would close the modal?

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 13 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:112]: I believe we still want to clear the input value state, to default to show the placeholder next time the modal is opened. Keep in mind this component is not unmounted, so we need to manage the state hook lifecycle.

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 11 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:68]: > The text box updates customSpendingCapInputValue but customSpendingCap is only set when submit is clicked?

> If that's the case, would it not be simpler to only keep customSpendingCapInputValue as state.

`customSpendingCapInputValue` is the value on the text input inside the modal. It needs to be a positive number.

`customSpendingCap` is set when the save button on the modal is pressed. This is the spending cap shown in the confirmation.

Screenshot 2024-09-10 at 10 29 09

> And then do any validation and dispatch imperatively inside the handleSubmit rather than during render via a useEffect?

Indeed, now that I moved a lot more of the logic to the modal component, it makes sense to move the dispatch inside the handleSubmit directly 👍 

The validation needs to be tied to the `customSpendingCap` to avoid showing a negative number in the confirmation screen (although the user can type it in the modal input as there's no way to limit at that level currently).

In summary:

customSpendingCapInputValue (Modal Input) -> Press "Save" -> coerce to positive number -> customSpendingCap (Confirmation) -> Press "Confirm" -> dispatch

**Review by jpuri — DISMISSED** [review]

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 14 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:151]: Again, submit closes the modal via the state hook but it doesn't unmount the component nor does it wipe other states.

**Review by jpuri — DISMISSED** [review]

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 14 above)
Comment (matthewwalsh0) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:156]: But even if it's not unmounting, does it not make the lifecycle simpler if we start from the same place each time?

Why do we need to persist anything from the last usage?

Can we not always use the formatted value?

Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:156]: @matthewwalsh0 That would make this line simpler, but I don't think it's the expected solution from a user's perspective. Let's consider this scenario:

The dApp suggested spending cap is 3 DAI for example, and the user edits the spending cap to 5 DAI using the modal. He then opens the modal again to edit the spending cap to 6 DAI.

The first time the modal is opened we want to show 3 DAI as a placeholder, while the second time we want to show 5 DAI, not 3.

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 11 above)
Comment (matthewwalsh0) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:68]: We can still generate the required custom data via the hook based on the input and reference those values within the callback. But if the `dispatch` is fundamentally only done on submit, then would it not be more explicit and more readable to actually call `dispatch` from within `handleSubmit` rather than a `useEffect`?

My confusion is why `customSpendingCap` needs to be state at all, can it not be derived in a standard variable and then rendered?

Why does it need to be persisted or generated at all if we hide the modal on submit?

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 14 above)
Comment (matthewwalsh0) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:156]: Exactly, but we've updated the transaction in the controller state, so won't the `formattedSpendingCap` automatically reflect the latest value as it's parsed from the transaction metadata which is now different?

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 11 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:68]: > We can still generate the required custom data via the hook based on the input and reference those values within the callback. But if the dispatch is fundamentally only done on submit, then would it not be more explicit and more readable to actually call dispatch from within handleSubmit rather than a useEffect?

I tried this approach but ran into a limitation due to needing to wait for the async hook that calculates the gas limit.

I worked around it by polling inside handleSubmit until the gas limit was calculated, but that is fundamentally flawed because the user can then confirm the confirmation before the dispatch is sent, resulting in a failure or inconsistent state. Here's the code I tested:

```tsx
 const { estimatedGasLimit } = useTransactionGasEstimate(
    transactionMeta,
    customTxParamsData,
    customSpendingCapInputValue,
  );

  const [isGasEstimateReady, setIsGasEstimateReady] = useState(false);

  useEffect(() => {
    if (estimatedGasLimit) {
      setIsGasEstimateReady(true);
    } else {
      setIsGasEstimateReady(false);
    }
  }, [estimatedGasLimit]);

  const handleCancel = useCallback(() => {
    setIsOpenEditSpendingCapModal(false);
    setCustomSpendingCapInputValue('');
    setIsGasEstimateReady(false);
  }, [setIsOpenEditSpendingCapModal, setCustomSpendingCapInputValue]);

  const waitForCondition = async (
    condition: boolean,
    interval: number,
    callback: () => void,
  ): Promise<void> => {
    return new Promise((resolve, reject) => {
      const checkCondition = async () => {
        try {
          if (condition) {
            callback();
            resolve();
          } else {
            setTimeout(checkCondition, interval);
          }
        } catch (error) {
          reject(error);
        }
      };

      checkCondition();
    });
  };

  const handleSubmit = useCallback(async () => {
    await waitForCondition(!isGasEstimateReady, 150, () => {
      dispatch(
        updateEditableParams(transactionMeta.id, {
          data: customTxParamsData,
          gas: hexToDecimal(estimatedGasLimit as string),
        }),
      );

      setIsOpenEditSpendingCapModal(false);
      setCustomSpendingCapInputValue('');
    });
  }, [
    isGasEstimateReady,
    customTxParamsData,
    estimatedGasLimit,
    setIsOpenEditSpendingCapModal,
    setCustomSpendingCapInputValue,
  ]);
```

I still believe the best solution is using the useEffect.

> My other confusion is why customSpendingCap needs to be state at all, can it not be derived in a standard variable and then rendered?

customSpendingCap functions as an intermediate state between the input field (setCustomSpendingCapInputValue) and the decoded spending cap inside the transactionMeta. It is used for displaying the value on the confirmation, to coerce negative input values into a spending cap of "0". This intermediate value is also important to trigger the dispatch, and without it we would have to dispatch every time the user types on the input field, which is an overkill in this case.

> Why does it need to be persisted or generated at all if we hide the modal on submit?

As per the comment below (https://github.com/MetaMask/metamask-extension/pull/26845#discussion_r1751628982), when the modal hides the component is not unmounted.

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 14 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:156]: I see your point now, and I agree. Pushed the changes 👍

File: test/e2e/tests/confirmations/transactions/increase-token-allowance-redesign.spec.ts [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+import { scrollAndConfirmAndAssertConfirm } from '../helpers';
+import { openDAppWithContract, TestSuiteArguments } from './shared';
+
+const {
+  defaultGanacheOptions,
+  defaultGanacheOptionsForType2Transactions,
+  withFixtures,
+} = require('../../../helpers');
+const FixtureBuilder = require('../../../fixture-builder');
+const { SMART_CONTRACTS } = require('../../../seeder/smart-contracts');
```
Comment (digiwand) [inline test/e2e/tests/confirmations/transactions/increase-token-allowance-redesign.spec.ts:18]:
```suggestion
import FixtureBuilder from '../../../fixture-builder';
import {
  defaultGanacheOptions,
  defaultGanacheOptionsForType2Transactions,
  largeDelayMs,
  veryLargeDelayMs,
  WINDOW_TITLES,
  withFixtures,
} from '../../../helpers';
import { Mockttp } from '../../../mock-e2e';
import { SMART_CONTRACTS } from '../../../seeder/smart-contracts';
import { Driver } from '../../../webdriver/driver';
import { scrollAndConfirmAndAssertConfirm } from '../helpers';
import { openDAppWithContract, TestSuiteArguments } from './shared';
```

I think we can remove the eslint-disable line and combine the `'../../../helpers'` imports

File: test/e2e/tests/confirmations/transactions/increase-token-allowance-redesign.spec.ts [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+          await openDAppWithContract(driver, contractRegistry, smartContract);
+
+          await createERC20IncreaseAllowanceTransaction(driver);
+
+          const NEW_SPENDING_CAP = '3';
+          await editSpendingCap(driver, NEW_SPENDING_CAP);
+
+          await scrollAndConfirmAndAssertConfirm(driver);
+
+          await assertChangedSpendingCap(driver, NEW_SPENDING_CAP);
```
Comment (digiwand) [inline test/e2e/tests/confirmations/transactions/increase-token-allowance-redesign.spec.ts:52]: The logic appears the same across these four tests except with varying spending caps. What if we move these into a reusable function? e.g. `createAndAssertIncreaseAllowanceSubmission(updateSpendingCap)`

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx (hunk 11 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx]: I added the dispatch and the calls to the util methods directly to the handleSubmit method as discussed. I also added a loading indicator to the save button.

File: test/e2e/tests/confirmations/transactions/increase-token-allowance-redesign.spec.ts [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+            .build(),
+          ganacheOptions: defaultGanacheOptionsForType2Transactions,
+          smartContract,
+          testSpecificMock: mocks,
+          title: this.test?.fullTitle(),
+        },
+        async ({ driver, contractRegistry }: TestSuiteArguments) => {
+          await openDAppWithContract(driver, contractRegistry, smartContract);
+
+          await createAndAssertIncreaseAllowanceSubmission(driver, '3000');
```
Comment (digiwand) [inline test/e2e/tests/confirmations/transactions/increase-token-allowance-redesign.spec.ts:119]: [suggestion] we could combine openDAppWithContract into createAndAssertIncreaseAllowanceSubmission where
createAndAssertIncreaseAllowanceSubmission would accept 3 params: driver, contractRegistry, newSpendingCap

File: test/e2e/webdriver/driver.js [hunk 18]
```diff
@@ -589,6 +589,17 @@ class Driver {
     }
   }
 
+  async takeScreenshot(title) {
+    const artifactDir = `./test-artifacts/${this.browser}/${title}`;
```
Comment (digiwand) [inline test/e2e/webdriver/driver.js:593]: [nit] we can move this as a const at the top of the file since it's being used in another area of this test

nice method!

File: test/e2e/webdriver/driver.js [hunk 19]
```diff
@@ -589,6 +589,17 @@ class Driver {
     }
   }
 
+  async takeScreenshot(title) {
```
Comment (digiwand) [inline test/e2e/webdriver/driver.js:592]:
```suggestion
  /** @param {string} title - The title of the window or tab the screenshot is being taken in */
  async takeScreenshot(title) {
```

File: test/e2e/tests/confirmations/transactions/increase-token-allowance-redesign.spec.ts [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+                redesignedConfirmationsEnabled: true,
+                isRedesignedConfirmationsDeveloperEnabled: true,
+              },
+            })
+            .build(),
+          ganacheOptions: defaultGanacheOptions,
+          smartContract,
+          testSpecificMock: mocks,
+          title: this.test?.fullTitle(),
+        },
```
Comment (digiwand) [inline test/e2e/tests/confirmations/transactions/increase-token-allowance-redesign.spec.ts:38]: [suggestion] I think we could DRY here more by creating a function to return the fixture options

```suggestion
        generateFixtureOptions({ ganacheOptions: isRedesignedConfirmationsDeveloperEnabled})
```

where the function could be something like
```
function createFixtureOptions(options = {}): object {
  return {
    dapp: true,
    fixtures: new FixtureBuilder()
      .withPermissionControllerConnectedToTestDapp()
      .withPreferencesController({
        preferences: {
          redesignedConfirmationsEnabled: true,
          isRedesignedConfirmationsDeveloperEnabled: true,
        },
      })
      .build(),
    ganacheOptions: defaultGanacheOptions,
    smartContract,
    testSpecificMock: mocks,
    title: this.test?.fullTitle(),
    ...options,
  };
}
```

File: ui/pages/confirmations/utils/confirm.ts [hunk 21]
```diff
@@ -24,6 +24,7 @@ export const REDESIGN_USER_TRANSACTION_TYPES = [
 export const REDESIGN_DEV_TRANSACTION_TYPES = [
```
Comment (digiwand) [inline ui/pages/confirmations/utils/confirm.ts:24]: Q: what does DEV represent in REDESIGN_DEV_TRANSACTION_TYPES?

**Review by matthewwalsh0 — DISMISSED** [review]

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+} from '../../../../../../../store/actions';
+import { getCustomTxParamsData } from '../../../../../confirm-approve/confirm-approve.util';
+import { useConfirmContext } from '../../../../../context/confirm';
+import { useAssetDetails } from '../../../../../hooks/useAssetDetails';
+import { useApproveTokenSimulation } from '../hooks/use-approve-token-simulation';
+
+export const getAccountBalance = (userBalance: string, decimals: string) =>
+  new BigNumber(userBalance)
+    .dividedBy(new BigNumber(10).pow(Number(decimals)))
+    .toNumber();
```
Comment (digiwand) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:38]: we can use `calcTokenAmount` where `getAccountBalance` is called.

we should also keep the value as a string instead of a number because converting to numbers may result in a loss of precision when the numbers are greater than Number.MAX_SAFE_INTEGER

. e.g.

```
> const bn = new BigNumber('1000000000000000128');
> bn.toFixed(): '1000000000000000128'
> bn.toNumber():  1000000000000000100
```

**Review by digiwand — CHANGES_REQUESTED** [review]
Requesting to update the accountBalance logic to keep the value as a string to avoid loss of precision

File: ui/pages/confirmations/components/confirm/info/approve/hooks/use-approve-token-simulation.ts [hunk 23]
```diff
@@ -18,34 +18,35 @@ export const useApproveTokenSimulation = (
...
-    tokenAmount = UNLIMITED_MSG;
-  } else {
-    tokenAmount = `${tokenPrefix}${formattedTokenNum}`;
-  }
+  const decodedSpendingCap = useMemo(() => {
+    return value
+      ? new BigNumber(value.data[0].params[1].value)
+          .dividedBy(new BigNumber(10).pow(Number(decimals)))
+          .toNumber()
+      : 0;
```
Comment (digiwand) [inline ui/pages/confirmations/components/confirm/info/approve/hooks/use-approve-token-simulation.ts:30]: similar to https://github.com/MetaMask/metamask-extension/pull/26845#discussion_r1754570299
 
 we can use `calcTokenAmount(value.data[0].params[1].value, decimals).toFixed()` here
```suggestion
    return value
      ? calcTokenAmount(value.data[0].params[1].value, decimals).toFixed()
      : 0;
```

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+
+  const { userBalance, tokenSymbol, decimals } = useAssetDetails(
+    transactionMeta.txParams.to,
+    transactionMeta.txParams.from,
+    transactionMeta.txParams.data,
+  );
+
+  const accountBalance = Number(
+    calcTokenAmount(userBalance || '0', Number(decimals || '0')),
+  );
```
Comment (digiwand) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:58]:
```suggestion
  const accountBalance = calcTokenAmount(userBalance ?? '0', decimals ?? '0').toFixed();
```

similar to the comment https://github.com/MetaMask/metamask-extension/pull/26845#discussion_r1754570299, we want to avoid transforming the value into a number to preserve the precision. Instead we can use `.toFixed()`. This can be applied to the other accountBalance instances as well. 

for `decimals ?? '0'`, It might be a good idea for us to do this within the method given we are doing this for a few other uses in this PR

File: ui/pages/confirmations/constants/index.ts [hunk 25]
```diff
@@ -7,3 +7,5 @@ export const TYPED_SIGNATURE_VERSIONS = {
   V3: 'V3',
   V4: 'V4',
 };
+
+export const UNLIMITED_MSG = 'UNLIMITED MESSAGE';
```
Comment (digiwand) [inline ui/pages/confirmations/constants/index.ts:11]: maybe SPENDING_CAP_UNLIMITED_MSG 

```suggestion
export const SPENDING_CAP_UNLIMITED_MSG = 'UNLIMITED MESSAGE';
```

File: ui/pages/confirmations/components/confirm/title/title.tsx [hunk 26]
```diff
@@ -102,6 +104,8 @@ const getDescription = (
         return t('confirmTitleDescApproveTransaction');
       }
       return t('confirmTitleDescERC20ApproveTransaction');
+    case TransactionType.tokenMethodIncreaseAllowance:
```
Comment (digiwand) [inline ui/pages/confirmations/components/confirm/title/title.tsx:107]: [nit] `TransactionType.tokenMethodApprove` and `TransactionType.tokenMethodIncreaseAllowance` cases can be moved to make the switch statement alphabetical. Same comment for the switch statement in getTitle above

File: ui/pages/confirmations/components/confirm/title/title.tsx [hunk 27]
```diff
@@ -102,6 +104,8 @@ const getDescription = (
...
+      return t('confirmTitleDescPermitSignature');
```
Comment (digiwand) [inline ui/pages/confirmations/components/confirm/title/title.tsx:108]: [suggestion] we can update these i18n keys since they are now being extended beyond signatures 

maybe
confirmTitleDescPermitSignature -> confirmTitleDescPermitTokens

**Review by digiwand — CHANGES_REQUESTED** [review]
For some reason, a related storybook page is not working
CleanShot 2024-09-13 at 00 52 47@2x
Is this the same behavior you are seeing? 

There may have also been a bug introduced prior to this PR to `develop` where the storybook app breaks after visiting certain storybook confirmation pages, and the application is no longer usable.

**Review by matthewwalsh0 — DISMISSED** [review]

**Review by matthewwalsh0 — DISMISSED** [review]

**Review by jpuri — DISMISSED** [review]

**Review by digiwand — COMMENTED** [review]
There are a couple of comments that I think got lost above. 2 of them involve precision concerns 

- https://github.com/MetaMask/metamask-extension/pull/26845/files#r1755165020
- https://github.com/MetaMask/metamask-extension/pull/26845/files#r1755158851

couple of other non-blocking, minor suggestions:
- https://github.com/MetaMask/metamask-extension/pull/26845/files#r1753941680
- https://github.com/MetaMask/metamask-extension/pull/26845/files#r1753899413
- https://github.com/MetaMask/metamask-extension/pull/26845/files#r1753887429

File: ui/pages/confirmations/components/confirm/info/approve/spending-cap/spending-cap.tsx [hunk 28]
```diff
@@ -91,9 +91,9 @@ export const SpendingCap = ({
   );
 
   const accountBalance = calcTokenAmount(
-    userBalance || '0',
-    Number(decimals || '0'),
-  );
+    userBalance ?? '0',
+    decimals,
```
Comment (digiwand) [inline ui/pages/confirmations/components/confirm/info/approve/spending-cap/spending-cap.tsx:95]:
```suggestion
    decimals ?? '0',
```
calcTokenAmount does not currently handle if decimals == null. Not sure if we should be concerned but maybe just in case?

File: ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx [hunk 29]
```diff
@@ -53,9 +53,10 @@ export const EditSpendingCapModal = ({
     transactionMeta.txParams.data,
   );
 
-  const accountBalance = Number(
-    calcTokenAmount(userBalance || '0', Number(decimals || '0')),
-  );
+  const accountBalance = calcTokenAmount(
+    userBalance ?? '0',
+    decimals,
```
Comment (digiwand) [inline ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx:58]:
```suggestion
    decimals ?? '0',
```
calcTokenAmount does not currently handle if decimals == null. Not sure if we should be concerned but maybe just in case?

Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx]: I added the Number coercion as well because calcTokenAmount expects decimals as a number

**Comment by digiwand** [comment]
Nice precision errors have been fixed! we will need to update the snapshots
CleanShot 2024-09-13 at 18 17 14@2x

**Review by digiwand — APPROVED** [review]

**Review by jpuri — APPROVED** [review]
